src/main/java/com/google/sps/servlets/ReviewsServlet.java

```diff
@@ -0,0 +1,122 @@
+    int rating = request.getParameter("rating");
+
+    // Get current timestamp for time when review is posted.
+    LocalDateTime dateTime = LocalDateTime.now();
+
+    // Get existing reviews key list.
+    List<Key> reviewsKeyList = getReviewsKeyList();
+
+    // Create review entity and get its key.
+    Entity reviewEntity = new Entity("Review");
+    reviewEntity.setProperty("userID", userID);
+    reviewEntity.setProperty("comment", comment);
+    reviewEntity.setProperty("rating", rating);
+    reviewEntity.setProperty("dateTime", dateTime);
+    datastore.put(reviewEntity);
+    Key reviewKey = reviewEntity.getKey();
+
+    // Append key to its Business entity.
+    reviewsKeyList.add(reviewKey);
+    businessEntity.setProperty(BUSINESS_REVIEWS, gson.toJson(reviewsKeyList));
+
+    // Redirect back to the product page.
+    response.sendRedirect("/product.html?businessID="+businessKey);
+  }
+
+  /**
+   * @return the String request parameter, or the default value if the parameter
+   *         was not specified by the client
+   */
+  private String getParameter(HttpServletRequest request, String name, String defaultValue) {
+    String value = request.getParameter(name);
+    if (value.isEmpty()) {
+      return defaultValue;
+    }
+    return value;
+  }
+
+  /** Get reviews from Datastore and send them as a Json to reviews.js. */
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Query Datastore for reviews under the appropriate Business entity.
+    List<Key> reviewsKeyList = getReviewsKeyList();
+    ArrayList<Review> reviews = new ArrayList<Review>();
+
+    // Convert review entities to objects and add to the list.
+    for (Key reviewKey : reviewsKeyList) {
+      Entity reviewEntity = datastore.get(reviewKey);
+      String userID = (String) reviewEntity.getProperty("userID");
+      String comment = (String) reviewEntity.getProperty("comment");
+      LocalDateTime dateTime = (String) reviewEntity.getProperty("dateTime");
+      int rating = reviewEntity.getProperty("rating");
+      Review review = new Review(userID, comment, rating, dateTime);
+      reviews.add(review);
+    }
+
+    String json = convertToJsonUsingGson(reviews);
+    response.setContentType("application/json;");
+    response.getWriter().println(json);
+  }
+
+  /**
+   * Converts a Java ArrayList<Review> into a JSON string using the Gson library.
+   */
+  private String convertToJsonUsingGson(ArrayList<Review> reviewsList) {
+    String json = gson.toJson(reviewsList);
+    return json;
+  }
+
+  /**
+   * Gets a list of Datastore keys corresponding to the Review entities for the Business.
+   * @return List<Key> of keys for review entities
+   */
+  private List<Key> getReviewsKeyList() {
+    String businessKey = request.getParameter(BUSINESS_KEY); // KIV url parameter
+    Entity businessEntity = datastore.get(businessKey);
```

Comment:
i think business key needs to be a Key type

Key businessKey =  KeyFactory.createKey("Business", idString);

Check line 97 onwards of my EditBusinessDataServlet.java for an example!

Reply:
`Key businessKey = KeyFactory.createKey("Business", Long.parseLong(request.getParameter(BUSINESS_KEY)));`
Ah I see. So this line creates a key for kind "Business", with the id being the URL parameter? Thanks Dalis!

Reply:
Yes!

Reply:
you should use "businessID" instead becuase the BUSINESS_KEY constant is specifically for my form and it's actually "businessId" Sorry for the inconsistency. I can go do one round of consistency checking for constants naming maybe after this week's sprint

Reply:
and also like Jireh mentioned below "businessID" could be something else depending on what Jacob uses